### PR TITLE
kv api of stroage

### PR DESCRIPTION
**Pull request by xiang90**
This pull requests implements the necessary storage kv api for the defined proto API.

Tnx operations are used to implement the Tnx service defined in the proto API.

### CODE REVIEW

**Comment by xiang90** [comment]
@yichengq @barakmich This is ready for review. The goal of this pull request is to set up a KV api and have a "working" implementation.

File: storage/kv.go [hunk 1]
```diff
@@ -1,132 +1,35 @@
...
-	return s.kvindex.Tombstone(key, currentIndex)
+import "github.com/coreos/etcd/storage/storagepb"
+
+type KV interface {
+	// Range gets the keys in the range at rangeIndex.
+	// If rangeIndex <=0, range gets the keys at currentIndex.
+	// If `end` is nil, the request returns the key.
+	// If `end` is not nil, it gets the keys in range [key, range_end).
+	// Limit limits the number of keys returned.
+	Range(key, end []byte, limit, rangeIndex int64) (kvs []storagepb.KeyValue, index int64)
```
Comment (yichengq) [inline storage/kv.go]: If the rangeIndex is compacted, it will return nothing. Do we expect that the upper level will take care of this case?

Comment (xiang90) [inline storage/kv.go]: There is no compaction right now. I have implemented compaction in my follow-up pr.

Comment (xiang90) [inline storage/kv.go]: if you want to see it: https://github.com/coreos/etcd/pull/2894

File: storage/kvstore.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	s.put(key, value, s.currentIndex+1)
+	s.TnxEnd(id)
+
+	return int64(s.currentIndex)
+}
+
+func (s *store) Range(key, end []byte, limit, rangeIndex int64) (kvs []storagepb.KeyValue, index int64) {
+	id := s.TnxBegin()
+	kvs, index = s.rangeKeys(key, end, limit, rangeIndex)
+	s.TnxEnd(id)
```
Comment (yichengq) [inline storage/kvstore.go]: maybe we could not use tnx for Range in the future? It looks like that we do some index logic only for this case, and we will support `Range` and `Put/Delete` happen at the same time.

**Comment by yichengq** [comment]
lgtm. the architecture and code is clean. I will jump in to write some related tests.

File: storage/kvstore.go (hunk 2 above)
Comment (xiang90) [inline storage/kvstore.go]: Yes. Probably we should just use the Rlock.

**Pull request merged**